hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java

```diff
@@ -746,66 +738,158 @@ private void addEventToQueryCache(Data dataKey, Record record) {
-        }
-        markRecordStoreExpirable(ttl, UNSET);
-        return true;
+        Object oldValue = putInternal(key, null, ttl, UNSET, getNow(),
+                true, true, false, true, false,
+                false, null, true, true,
+                null, null, true, backup);
+        return oldValue != null;
     }
 
     @Override
     public Object set(Data dataKey, Object value, long ttl, long maxIdle) {
-        return putInternal(dataKey, value, ttl, maxIdle, null, false, true);
+        return putInternal(dataKey, value, ttl, maxIdle, getNow(),
+                false, true, false, false, false,
+                false, null, false, true,
+                null, null, true, false);
     }
 
     @Override
     public Object setTxn(Data dataKey, Object value, long ttl, long maxIdle, UUID transactionId) {
-        return putInternal(dataKey, value, ttl, maxIdle, transactionId, false, true);
+        return putInternal(dataKey, value, ttl, maxIdle, getNow(),
+                false, true, false, false, false,
+                false, null, false, true,
+                transactionId, null, true, false);
     }
 
     @Override
     public Object put(Data key, Object value, long ttl, long maxIdle) {
-        return putInternal(key, value, ttl, maxIdle, null, true, true);
+        return putInternal(key, value, ttl, maxIdle, getNow(),
+                true, true, false, false, false,
+                false, null, false, true,
+                null, null, true, false);
     }
 
-    protected Object putInternal(Data key, Object newValue, long ttl,
-                                 long maxIdle, @Nullable UUID transactionId,
-                                 boolean loadFromStore, boolean countAsAccess) {
-        checkIfLoaded();
-
-        long now = getNow();
-        markRecordStoreExpirable(ttl, maxIdle);
-
-        Record record = getRecordOrNull(key, now, false);
-        Object oldValue = record == null
-                ? (loadFromStore ? loadValueOf(key) : null) : record.getValue();
-        newValue = mapServiceContext.interceptPut(interceptorRegistry, oldValue, newValue);
-        onStore(record);
+    /**
+     * Core put method for all variants of puts/updates.
+     * @return old value if this is an update operations, otherwise returns null
+     */
+    @SuppressWarnings({"checkstyle:npathcomplexity",
+            "checkstyle:parameternumber", "checkstyle:cyclomaticcomplexity"})
+    private Object putInternal(Data key, Object newValue, long ttl,
+                               long maxIdle, long now, boolean load, boolean store,
+                               boolean putIfAbsent, boolean putIfExists, boolean putIfEqual,
+                               boolean putFromLoad, Object expectedValue, boolean useOldValueIfNoNewValue,
+                               boolean checkIfLoaded, @Nullable UUID transactionId,
+                               Address callerAddress, boolean countAsAccess, boolean backup) {
```

Comment:
Merging will help here to see logical problems, fixing bugs or making improvements in one place for all variants of puts. For readability/understandability, i added short comments. For performance, i believe branching has very minimal effect, since we have still a lot of branching, but with this 2-3 new branches... it needs verifications of course. To see possible improvements  easily i'll keep the method as dumb as possible for now, so not planning to add PutControl.

Reply:
I agree that the performance loss is insignificant here. I commented that as well because the branches break the previously straight control flow similarly to how they break the "reading flow" (for me). Anyways, I'm not blocking you on this, we will see in the end if this unification helps or distracts. 

With the `PutControl` idea, I didn't mean any change to `putInternal()` itself. It would be only changing `if (putIfAbsent) {...}` to `if (ctrl.putIfAbsent) {...}`. It would be useful only to the non-IDE readers though, as the IDEs - at least IntelliJ - label the variables, so I'm ok with not implementing this as well.